src/core/ad.jl

```diff
@@ -12,7 +12,7 @@ Computes the gradient of the log joint of `θ` for the model specified by
 function gradient(
     θ::AbstractVector{<:Real},
     vi::VarInfo,
-    model::Function,
+    model,
```

Comment:
I think it would be better if this stays to be a function and not `model::CallableModel`. Or not?

Reply:
A `Function` is too restrictive. Julia lets us use [callable structs](https://docs.julialang.org/en/v1/manual/methods/index.html#Function-like-objects-1) which can store interesting information. The only purposes of input type annotations that I know of in Julia are: 1) to guide dispatch, 2) to throw meaningful errors for user-facing functions, or 3) to signal that a certain interface needs to be defined for the input, e.g. `AbstractVector`. I don't think `Function` is used for dispatch here, and the only "interface" assumed is that it needs to be callable so using `::Function` means more code and less features ;) Think of a user defining their own callable model struct, which can have a `supertype` other than `Function`, and then they want to sample from it. Not restricting the input here means that they can pass their callable struct through and it will just work, at least it should!